ch/server.go

```diff
@@ -0,0 +1,90 @@
+package ch
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"log"
+	"sync"
+	"time"
+
+	"github.com/ClickHouse/clickhouse-go/v2"
+)
+
+type (
+	ClickhouseProvider interface {
+		// FetchCredential retrieves clickhouse credential
+		FetchCredential(ctx context.Context, serverKey string) (ClickhouseCred, error)
+
+		// FetchConfig retrieve clickhouse configuration
+		FetchConfig(ctx context.Context, serverKey string) (ClickhouseConfig, error)
+	}
+
+	ClickhouseCred struct {
+		Protocol string `mapstructure:"protocol"`
+		Host     string `mapstructure:"host"`
+		HTTPPort string `mapstructure:"http_port"`
+		TCPPort  string `mapstructure:"tcp_port"`
+		Username string `mapstructure:"username"`
+		Password string `mapstructure:"password"`
+	}
+
+	ClickhouseConfig struct {
+		StoragePolicy       string `mapstructure:"storage_policy"`
+		MaxOpenConns        int    `mapstructure:"max_open_conns"`
+		MaxIdleConns        int    `mapstructure:"max_idle_conns"`
+		ConnMaxLifetimeSecs int    `mapstructure:"conn_max_lifetime"`
+	}
+
+	ClickhouseServer struct {
+		Credential ClickhouseCred   `mapstructure:"credential"`
+		Config     ClickhouseConfig `mapstructure:"config"`
+
+		clickhouse.Conn
+		mu sync.Mutex
+	}
+
+	ClickhouseServers struct {
+		Servers map[string]ClickhouseServer `mapstructure:"servers"`
+		mu      sync.Mutex
```

Comment:
Same as above - does not look like this is used.

Reply:
Exclusion lock. This is currently not used. I will remove it if necessary.